internal/operator/uninstall.go

```diff
@@ -0,0 +1,153 @@
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/yaml"
+)
+
+type Uninstall struct {
+	config *Configuration
+
+	Package string
+}
+
+func NewUninstall(cfg *Configuration) *Uninstall {
+	return &Uninstall{
+		config: cfg,
+	}
+}
+
+func (u *Uninstall) Run(ctx context.Context) error {
+	subs := v1alpha1.SubscriptionList{}
+	if err := u.config.Client.List(ctx, &subs, client.InNamespace(u.config.Namespace)); err != nil {
+		return fmt.Errorf("list subscriptions: %v", err)
+	}
+
+	var sub *v1alpha1.Subscription
+	for _, s := range subs.Items {
+		if u.Package == s.Spec.Package {
+			sub = &s
+			break
+		}
+	}
+	if sub == nil {
+		return fmt.Errorf("operator package %q not found", u.Package)
+	}
+
+	catsrcKey := types.NamespacedName{
+		Namespace: sub.Spec.CatalogSourceNamespace,
+		Name:      sub.Spec.CatalogSource,
+	}
+	catsrc := &v1alpha1.CatalogSource{}
+	if err := u.config.Client.Get(ctx, catsrcKey, catsrc); err != nil {
+		return fmt.Errorf("get catalog source: %v", err)
+	}
+
+	installPlanKey := types.NamespacedName{
+		Namespace: sub.Status.InstallPlanRef.Namespace,
+		Name:      sub.Status.InstallPlanRef.Name,
+	}
+
+	// Since the install plan is owned by the subscription, we need to
+	// read all of the resource references from the install plan before
+	// deleting the subscription.
+	deleteObjs, err := u.getInstallPlanResources(ctx, installPlanKey)
+	if err != nil {
+		return err
+	}
+
+	// Delete the subscription first, so that no further installs or upgrades
+	// of the operator occur while we're cleaning up.
+	if err := u.config.Client.Delete(ctx, sub); err != nil {
+		return fmt.Errorf("delete subscription %q: %v", sub.Name, err)
+	}
+	fmt.Printf("subscription %q deleted\n", sub.Name)
+
+	// Ensure CustomResourceDefinitions are deleted first, so that the operator
+	// has a chance to handle CRs that have finalizers.
+	sort.SliceStable(deleteObjs, func(i, j int) bool {
+		return deleteObjs[i].GetObjectKind().GroupVersionKind().Kind == "CustomResourceDefinition"
+	})
+	for _, obj := range deleteObjs {
+		err := u.config.Client.Delete(ctx, obj)
+		if err != nil && !apierrors.IsNotFound(err) {
+			return err
+		}
+		if !apierrors.IsNotFound(err) {
+			fmt.Printf("%s %q deleted\n", strings.ToLower(obj.GetObjectKind().GroupVersionKind().Kind), obj.GetName())
+		}
+	}
+
+	// Delete the catalog source. This assumes that all underlying resources related
+	// to this catalog source have an owner reference to this catalog source so that
+	// they are automatically garbage-collected.
+	if err := u.config.Client.Delete(ctx, catsrc); err != nil {
+		return fmt.Errorf("delete catalog source: %v", err)
+	}
+	fmt.Printf("catalogsource %q deleted\n", catsrc.Name)
+
+	// If this was the last subscription in the namespace, delete the operator group.
```

Comment:
Shouldn't this be optional?

Reply:
We _always_ deleted the operator group with `cleanup packagemanifests` right? So this is basically the same except we keep the operator group around if there's another subscription.

If someone asks for the ability to never delete the operator group, we can add a new flag after 1.0?

Reply:
`cleanup packagemanifests` always deletes the SDK-managed operator group [`operator-sdk-og`](https://github.com/operator-framework/operator-sdk/blob/ce96b751a9e377e4e138e76356bf2409ac1fc457/internal/olm/operator/olm.go#L154), whereas this code could delete any operator group in the namespace. This is unlikely to be an operator group other than the `operator-sdk-og` created by `run packagemanifests`, but is still possible. How about checking the operator group name before deleting to make sure it is the right one?

Reply:
Ah I see. Yeah, that works. We would want to make the eventual `run bundle` command use the same name for the operator group then.

@rashmigottipati @bharathi-tenneti @jmrodri FYI